src/transformers/models/convnext/modeling_tf_convnext.py

```diff
@@ -0,0 +1,609 @@
+        )
+        self.pwconv1 = tf.keras.layers.Dense(
+            units=4 * dim,
+            kernel_initializer=get_initializer(config.initializer_range),
+            bias_initializer="zeros",
+            name="pwconv1",
+        )  # pointwise/1x1 convs, implemented with linear layers
+        self.act = get_tf_activation(config.hidden_act)
+        self.pwconv2 = tf.keras.layers.Dense(
+            units=dim,
+            kernel_initializer=get_initializer(config.initializer_range),
+            bias_initializer="zeros",
+            name="pwconv2",
+        )
+        # Using `layers.Activation` instead of `tf.identity` to better control `training`
+        # behaviour.
+        self.drop_path = (
+            TFConvNextDropPath(
+                drop_path,
+                name="drop_path",
+            )
+            if drop_path > 0.0
+            else tf.keras.layers.Activation(
+                "linear",
+                name="drop_path",
+            )
+        )
+
+    def build(self, input_shape: tf.TensorShape):
+        # PT's `nn.Parameters` must be mapped to a TF layer weight to inherit the same name hierarchy (and vice-versa)
+        self.layer_scale_parameter = (
+            self.add_weight(
+                shape=(self.dim,),
+                initializer=tf.keras.initializers.Constant(value=self.config.layer_scale_init_value),
+                trainable=True,
+                name="layer_scale_parameter",
+            )
+            if self.config.layer_scale_init_value > 0
+            else None
+        )
+        super().build(input_shape)
+
+    def call(self, hidden_states, training=False):
+        input = hidden_states
+        x = self.dwconv(hidden_states)
+        x = self.layernorm(x)
+        x = self.pwconv1(x)
+        x = self.act(x)
+        x = self.pwconv2(x)
+
+        if self.layer_scale_parameter is not None:
+            x = self.layer_scale_parameter * x
+
+        x = input + self.drop_path(x, training=training)
+        return x
+
+
+class TFConvNextStage(tf.keras.layers.Layer):
+    """ConvNext stage, consisting of an optional downsampling layer + multiple residual blocks.
+
+    Args:
+        config ([`ConvNextConfig`]): Model configuration class.
+        in_channels (`int`): Number of input channels.
+        out_channels (`int`): Number of output channels.
+        depth (`int`): Number of residual blocks.
+        drop_path_rates(`List[float]`): Stochastic depth rates for each layer.
+    """
+
+    def __init__(
+        self, config, in_channels, out_channels, kernel_size=2, stride=2, depth=2, drop_path_rates=None, **kwargs
+    ):
+        super().__init__(**kwargs)
+        if in_channels != out_channels or stride > 1:
+            self.downsampling_layer = [
+                tf.keras.layers.LayerNormalization(
+                    epsilon=1e-6,
+                    name="downsampling_layer.0",
+                ),
+                tf.keras.layers.Conv2D(
```

Comment:
It would be very nice if there is a comment here mentioning what format we expect: NCHW or NHWC (with a short description) :-)

similar to [this place](https://github.com/huggingface/transformers/blob/15c916f416dfc5c820528f25c0257544847ceb50/src/transformers/models/convnext/modeling_tf_convnext.py#L103)

Reply:
The `data_format` in `layers.Conv2D` defaults to `channels_last` I think it's okay to keep it this way.

Reply:
Hi, I actually mean a short description about why the input would be `NHWC` at that point.

However, reading more code, I understand:

[TFConvNextEmbeddings](TFConvNextEmbeddings) change input from `NCHW` to `NHWC`, and it keeps this format until the output where we change back to `NCHW`.

(I just think a remark will be easier to know this more quickly :-) )

Reply:
I will do that right away.

Reply:
See if the comment is helpful.

Reply:
Good! (And definitely very helpful for me, hopefully also for some other people too :-) ) Thanks!